### PR TITLE
Implement a Service Bus Shared Access Key Credential

### PR DESCRIPTION
**Pull request by v-hongli1**
Fix issue #16465 
1. Add and change class and test case for implement a Service Bus Shared Access Key Credential
2. Add and change files:
   1). sdk/identity/azure-identity/src/main/java/com/azure/identity/ServiceBusSharedKeyCredential.java
   2). sdk/identity/azure-identity/src/main/java/com/azure/identity/ServiceBusSharedKeyCredentialBuilder.java
   3). sdk/identity/azure-identity/src/test/java/com/azure/identity/ServiceBusSharedKeyCredentialTest.java
   4). sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/util/ValidationUtil.java
3. Add a new validation method `validateForSharedAccessKey` in the `ValidationUtil` class to cope with the two parameter setting methods when the Shared Access Key Credential is created. 
   a. When `sharedAccessSignature != null `, the specified `sharedAccessSignature` can be directly used to create Shared Access Key Credential, so `sharedAccessPolicy` and `shardAccessKey` do not require any verification.
b. When `sharedAccessSignature == null`, `sharedAccessPolicy` and `shardAccessKey` are required for verification. 

@jongio, @conniey for notification.

### CODE REVIEW

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/ServiceBusSharedKeyCredential.java [hunk 1]
```diff
@@ -0,0 +1,140 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
[the remark this answers is not in the document]
Comment (hongli750210) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/ServiceBusSharedKeyCredential.java]: According to your comment, make the following adjustments in the new fix version:
1. `RemoveServiceBusSharedKeyCredential` class and related builder class and test case from azure-identity.
2. The client builders in Service Bus support `AzureNamedKeyCredential` and `AzureSasCredential` classes as overloads to the existing `credential()` builder.

File: sdk/identity/azure-identity/src/main/java/module-info.java [hunk 2]
```diff
@@ -12,6 +12,7 @@
     requires org.linguafranca.pwdb.database;
     requires org.linguafranca.pwdb.kdbx;
     requires org.linguafranca.pwdb.kdbx.simple;
+    requires reactor.core;
```
Comment (conniey) [inline sdk/identity/azure-identity/src/main/java/module-info.java:15]: This change should be reverted.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusClientBuilder.java [hunk 3]
```diff
@@ -7,18 +7,20 @@
...
 import com.azure.core.amqp.ProxyOptions;
-import com.azure.core.amqp.implementation.AzureTokenManagerProvider;
-import com.azure.core.amqp.implementation.CbsAuthorizationType;
-import com.azure.core.amqp.implementation.ConnectionOptions;
-import com.azure.core.amqp.implementation.ConnectionStringProperties;
 import com.azure.core.amqp.implementation.MessageSerializer;
-import com.azure.core.amqp.implementation.ReactorHandlerProvider;
-import com.azure.core.amqp.implementation.ReactorProvider;
+import com.azure.core.amqp.implementation.TracerProvider;
+import com.azure.core.amqp.implementation.ConnectionStringProperties;
```
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusClientBuilder.java:12]: This should be reordered. IntelliJ has a keyboard shortcut you can use too:
https://www.jetbrains.com/help/idea/creating-and-optimizing-imports.html#optimize-imports

File: sdk/identity/azure-identity/src/main/java/module-info.java (hunk 2 above)
Comment (hongli750210) [reply sdk/identity/azure-identity/src/main/java/module-info.java:15]: According to your comment in the new revision amendments have been made.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusClientBuilder.java (hunk 3 above)
Comment (hongli750210) [reply sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusClientBuilder.java:12]: According to your comment in the new revision amendments have been made.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusClientBuilder.java [hunk 4]
```diff
@@ -235,6 +237,55 @@ public ServiceBusClientBuilder credential(String fullyQualifiedNamespace, TokenC
         return this;
     }
 
+    /**
+     * Sets the credential for the Service Bus resource.
+     *
+     * @param fullyQualifiedNamespace for the Service Bus.
+     * @param credential {@link TokenCredential} to be used for authentication.
```
Comment (ramya-rao-a) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusClientBuilder.java:244]: This overload does not use TokenCredential, so this comment should be updated. Same for the other overload that takes the AzureSasCredential

Comment (hongli750210) [reply sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusClientBuilder.java:244]: Sorry, I missed the correction of comment. According to your comment in the new revision amendments have been made.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusSenderAsyncClient.java [hunk 5]
```diff
@@ -47,10 +47,7 @@
 import static com.azure.core.amqp.implementation.RetryUtil.withRetry;
 import static com.azure.core.util.FluxUtil.fluxError;
 import static com.azure.core.util.FluxUtil.monoError;
-import static com.azure.core.util.tracing.Tracer.AZ_TRACING_NAMESPACE_KEY;
-import static com.azure.core.util.tracing.Tracer.ENTITY_PATH_KEY;
-import static com.azure.core.util.tracing.Tracer.HOST_NAME_KEY;
-import static com.azure.core.util.tracing.Tracer.SPAN_CONTEXT_KEY;
+import static com.azure.core.util.tracing.Tracer.*;
```
Comment (conniey) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusSenderAsyncClient.java:50]: Please update your settings not to use * import.

https://stackoverflow.com/a/3589885/4220757

**Review by conniey — COMMENTED** [review]
Looks good aside from the CI build failures.

File: sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusClientBuilderTest.java [hunk 6]
```diff
@@ -243,6 +256,103 @@ public void testConnectionStringWithSas() {
...
+                })
+            ).verifyComplete();
+        } finally {
+            senderAsyncClient.close();
+        }
+    }
+
+
+    @Test
+    public void sendAndReceiveEventByAzureSasCredential() {
```
Comment (hemanttanwar) [inline sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusClientBuilderTest.java:288]: It says sendAndReceiveEvents but the test is only sending  and receiving.

**Review by hemanttanwar — COMMENTED** [review]
Could we add test in integration test for receiving or sending message using sas key ?

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusSenderAsyncClient.java (hunk 5 above)
Comment (hongli750210) [reply sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusSenderAsyncClient.java:50]: According to your Comment, fixed in new version.

File: sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusClientBuilderTest.java (hunk 6 above)
Comment (hongli750210) [reply sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusClientBuilderTest.java:288]: According to your Comment, change method name to `testBatchSendEventByAzureSasCredential`

**Comment by jongio** [comment]
@conniey / @hemanttanwar - Do you need anything else on this?

**Review by conniey — APPROVED** [review]

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusClientBuilder.java [hunk 7]
```diff
@@ -235,6 +237,55 @@ public ServiceBusClientBuilder credential(String fullyQualifiedNamespace, TokenC
...
+            "'fullyQualifiedNamespace' cannot be null.");
+        Objects.requireNonNull(credential, "'credential' cannot be null.");
+
+        this.credentials = new ServiceBusSharedKeyCredential(credential.getAzureNamedKey().getName(),
+            credential.getAzureNamedKey().getKey(), ServiceBusConstants.TOKEN_VALIDITY);
+
+        if (CoreUtils.isNullOrEmpty(fullyQualifiedNamespace)) {
+            throw logger.logExceptionAsError(
+                new IllegalArgumentException("'fullyQualifiedNamespace' cannot be an empty string."));
+        }
```
Comment (srnagar) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusClientBuilder.java:260]: All input validation should happen before performing any operations on the input. Here `ServiceBusSharedKeyCredential` should not be created until `fullyQualifiedNamespace` param is validated.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusClientBuilder.java [hunk 8]
```diff
@@ -235,6 +237,55 @@ public ServiceBusClientBuilder credential(String fullyQualifiedNamespace, TokenC
...
+     *
+     * @return The updated {@link ServiceBusClientBuilder} object.
+     */
+    public ServiceBusClientBuilder credential(String fullyQualifiedNamespace, AzureSasCredential credential) {
+
+        this.fullyQualifiedNamespace = Objects.requireNonNull(fullyQualifiedNamespace,
+            "'fullyQualifiedNamespace' cannot be null.");
+        Objects.requireNonNull(credential, "'credential' cannot be null.");
+
+        this.credentials = new ServiceBusSharedKeyCredential(credential.getSignature());
```
Comment (srnagar) [inline sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusClientBuilder.java:279]: This overwrites the value set in the previous overload. So, the order of calls to `credential()` decides which credential is being used. Instead, we should do the validate in `build` method and throw exception if both are set.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusClientBuilder.java [hunk 9]
```diff
@@ -235,6 +237,55 @@ public ServiceBusClientBuilder credential(String fullyQualifiedNamespace, TokenC
         return this;
     }
 
+    /**
+     * Sets the credential for the Service Bus resource.
```
[the remark this answers is not in the document]
Comment (hongli750210) [reply sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusClientBuilder.java:241]: According to your comment, fixed in the new version.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusClientBuilder.java (hunk 7 above)
Comment (hongli750210) [reply sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusClientBuilder.java:260]: According to your comment, fixed in the new version.

File: sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusClientBuilder.java (hunk 8 above)
Comment (hongli750210) [reply sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusClientBuilder.java:279]: According to your comment, fixed in the new version.

**Pull request merged**